2022/incoming/a-perlmas-tree.pod

```diff
@@ -0,0 +1,221 @@
+    $spinner->size_hint_align_set(EVAS_HINT_FILL,0.5);
+    $spinner->size_hint_weight_set(EVAS_HINT_EXPAND,0.0);
+    $spinner->min_max_set($opts->{min},$opts->{max});
+    $spinner->step_set(1);
+    $spinner->indicator_format_set("%1.0f");
+    $spinner->value_set($settings{$opts->{key}});
+    $spinner->show(); $table->pack($spinner,1,$row,2,1);
+
+    $spinner->smart_callback_add("delay,changed" => sub {$settings{$opts->{key}} = int($_[1]->value_get());}, undef);
+}
+
+=end perl
+    
+Most of the code should be self-explaining. Therefore only two little explanations: 
+
+With C<< $spinner->indicator_format_set("%1.0f"); >> the popup for the actual value shows only integers.
+
+Very important is the C<< $spinner->smart_callback_add("eventName",\&eventSub,$data) >>. With this you can register an event handler for smart callbacks, the most often used event types in EFL.The callback function takes as parameters the additional C<< $data >> defined in the C<< $widget->smart_callback_add() >>, the object to which the event happened, and the event info data, which depends on the object type and the event at play. So the signature is
+
+=begin perl
+
+sub smart_event_cb {
+    my ($data, $obj, $evInfo) = @_;
+}
+
+=end perl
+
+In the example code we use an anonymous subroutine as event handler, that sets the key in the setting hash with the actual value of the spinner object.
+
+=item 3. Last we add the button to create the Christmas tree to the container-box 
+
+and show the window (don't forget this!!)
+
+=begin perl
+
+my $btn = pEFL::Elm::Button->add($container);
+$btn->text_set("Create a new Christmas Tree");
+$btn->smart_callback_add("clicked",\&create_christmas_tree,$viewer);
+$btn->show(); $container->pack_end($btn);
+
+$win->show()
+
+=end perl
+        
+Again we register a callback to the follwing event handler that creates the SVG file and let the viewer show
+the created tree. A little hack is needed with C<< $viewer->file_set(undef,"") >>. Otherwise the viewer doesn't know that the file has changed:
+
+=begin perl
+
+sub create_christmas_tree {
+    my ($viewer,$obj,$evinfo) = @_;
+        
+    my $tree = SVG::ChristmasTree->new(\%settings);
+
+    open my $fh, ">", "tree.svg" or die "Could not write to ./tree.svg: $!\n";
+    print $fh $tree->as_xml;
+    close $fh;
+
+    $viewer->file_set(undef,"");
+    $viewer->file_set("./tree.svg","");
+}
+
+=end perl
+
+=back 
+    
+=head1 Conclusion
+
+I hope you enjoyed this little introduction to pEFL. Because the perl advent articles should be light and trivial, the explained version only gives you control over the numeric settings of SVG::ChristmasTree. But at L<my Perl Advent 2022 github repository|https://github.com/MaxPerl/PerlAdvent2022> you find also an advanced version where you can also alter the colors of the tree, baubles and pot. The additions adjust the colorselector examples under examples/colorselector.pl in the pEFL distribution and hopefully are understandable.
```

Comment:
```suggestion
I hope you enjoyed this little introduction to pEFL. Because the Perl advent articles should be light and trivial, the explained version only gives you control over the numeric settings of SVG::ChristmasTree. But at L<my Perl Advent 2022 github repository|https://github.com/MaxPerl/PerlAdvent2022> you find also an advanced version where you can also alter the colors of the tree, baubles and pot. The additions adjust the colorselector examples under examples/colorselector.pl in the pEFL distribution and hopefully are understandable.
```